flatgfa/src/packedseq.rs

```diff
@@ -0,0 +1,293 @@
+    high_nibble_end: bool,
+}
+
+impl PackedVec {
+    fn new() -> Self {
+        PackedVec {
+            data: Vec::new(),
+            high_nibble_end: false,
+        }
+    }
+    pub fn push(&mut self, value: u8) {
+
+        if self.high_nibble_end {
+            self.data.push(value);
+            self.high_nibble_end = false;
+        } else {
+            let last_index = self.data.len() - 1;
+            self.data[last_index] = (value << 4) | self.data[last_index];
+            self.high_nibble_end = true;
+        }
+    }
+}
+
+// impl Index<usize> for PackedVec {
+//     type Output = u8;
+//     fn index(&self, index: usize) -> &Self::Output {
+//         if index % 2 == 1 {
+//             (&self.data[index as usize / 2] & 0b11110000u8) >> 4
+//         } else {
+//             &self.data[index as usize / 2] & 0b00001111u8
+//         }
+//     }
+// }
+
+impl fmt::Display for PackedVec {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[")?;
+        let iter = PackedVecIterator {
+            data: &self.data,
+            cur_index: 0,
+            cur_high_nibble: false,
+            high_nibble_end: self.high_nibble_end,
+            last: false
+        };
+        let mut i = 0;
+        for item in iter {
+            if i == 0 {
+                write!(f, "{}", match item {0 => 'A', 1 => 'C', 2 => 'G', 3 => 'T', _ => ' '})?; 
+                i = 1; 
+            } else {
+            write!(f, ", {}",  match item {0 => 'A', 1 => 'C', 2 => 'G', 3 => 'T', _ => ' '})?;
+            }
+        }
+        write!(f, "]")
+    }
+}
+
+struct PackedVecIterator<'a> {
+    data: &'a Vec<u8>,
+    cur_index: usize,
+    cur_high_nibble: bool,
+    high_nibble_end: bool,
+    last: bool
+}
+
+impl<'a> Iterator for PackedVecIterator<'a> {
+
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cur_index >= self.data.len() || self.last {
+            return None; 
+        }
+        if self.cur_index == self.data.len() - 1 && !self.high_nibble_end {
+           self.last = true;
+        }
+        let result = if self.cur_high_nibble {
+            (self.data[self.cur_index] >> 4) & 0b00001111u8  
+        }
+        else {
+            self.data[self.cur_index]& 0b00001111u8  
+        };
+        if self.cur_high_nibble {
+            self.cur_index += 1;
+        }
+        self.cur_high_nibble = !self.cur_high_nibble;
+        Some(result)
+    }
+}
+
+/// A reference to a subsection of a nucleotide sequence stored in a PackedVec 
+/// data_ref: The underlying vector that stores the sequence referenced by this slice
+/// span: The specific section of the sequence that this slice references
+struct PackedSlice<'a> {
+    data_ref: &'a Vec<u8>,
+    span: Span
+}
+
+
+/// Used by PackedSlices to reference a subsection of a nucleotide sequence  
+struct Span {
+    start: u8,
+    end: u8,
+}
+
+
+fn get_seq(compressed_data: &Vec<u8> , span: Span)  -> Vec<u8> {
```

Comment:
For functions like this, I think it would make sense to move them to methods in the `impl PackedVec` implementation. This would make it easier to share functionality and would keep the utilities clearly organized around the data structure they work with. Do you think that seems feasible?